File: src/main/java/com/github/holly/accountability/config/auth/AccountabilityUserDetailsService.java
package com.github.holly.accountability.config.auth;

import com.github.holly.accountability.user.AccountabilitySessionUser;
import com.github.holly.accountability.user.User;
import com.github.holly.accountability.user.UserRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class AccountabilityUserDetailsService implements UserDetailsService {

    @Autowired
    private UserRepository userRepository;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {

        User databaseUser = userRepository
                .findByUsernameIgnoreCase(username)
                .orElseThrow(() -> new UsernameNotFoundException(username));

        return new AccountabilitySessionUser(
                databaseUser.getId(),
                databaseUser.getName(),
                databaseUser.getEmail(),
                databaseUser.getUsername(),
                databaseUser.getPassword(),
                List.of(new SimpleGrantedAuthority("ROLE_USER"))
        );
    }
}

File: src/main/kotlin/com/github/holly/accountability/config/auth/AccountabilityUserDetailsService.kt
package com.github.holly.accountability.config.auth

import com.github.holly.accountability.user.AccountabilitySessionUser
import com.github.holly.accountability.user.UserRepository
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

@Service
class AccountabilityUserDetailsService(
    val userRepository: UserRepository
): UserDetailsService {

    override fun loadUserByUsername(username: String?): UserDetails? {
        val databaseUser = userRepository.findByUsernameIgnoreCase(username!!) ?: throw UsernameNotFoundException(username)
        return AccountabilitySessionUser(databaseUser.id!!, databaseUser.username, databaseUser.name, databaseUser.password, listOf(SimpleGrantedAuthority("ROLE_USER")))
    }
}